src/System.Net.Http/tests/FunctionalTests/HttpCookieProtocolTests.cs

```diff
@@ -40,58 +40,58 @@ private static CookieContainer CreateSingleCookieContainer(Uri uri, string cooki
 
         private static string GetCookieHeaderValue(string cookieName, string cookieValue) => $"{cookieName}={cookieValue}";
 
-
-        [Fact]
-        public async Task GetAsync_DefaultCoookieContainer_NoCookieSent()
+        [Theory]
+        [MemberData(nameof(LoopbackServerFactoriesMemberData))]
+        public async Task GetAsync_DefaultCoookieContainer_NoCookieSent(LoopbackServerFactory serverFactory)
         {
-            await LoopbackServer.CreateServerAsync(async (server, url) =>
-            {
-                using (HttpClient client = CreateHttpClient())
+            await serverFactory.CreateClientAndServerAsync( 
+                async uri =>
                 {
-                    Task<HttpResponseMessage> getResponseTask = client.GetAsync(url);
-                    Task<List<string>> serverTask = server.AcceptConnectionSendResponseAndCloseAsync();
-                    await TestHelper.WhenAllCompletedOrAnyFailed(getResponseTask, serverTask);
-
-                    List<string> requestLines = await serverTask;
-
-                    Assert.Equal(0, requestLines.Count(s => s.StartsWith("Cookie:")));
-                }
-            });
+                    using (HttpClient client = CreateHttpClient(true))
+                    {
+                        await client.GetAsync(uri);
+                    }
+                },
+                async server =>
+                {
+                    HttpRequestData requestData = await server.HandleRequestAsync();
+                    Assert.Equal(0, requestData.GetHeaderValueCount("Cookie"));
+                });
         }
 
         [Theory]
-        [MemberData(nameof(CookieNamesValuesAndUseCookies))]
-        public async Task GetAsync_SetCookieContainer_CookieSent(string cookieName, string cookieValue, bool useCookies)
+        [MemberData(nameof(LoopbackServersAndCookieNamesValuesAndUseCookies))]
```

Comment:
Maybe we should introduce a new MemberData-like attribute that provides the cross-product of all of the inputs, e.g. so you could write:
```C#
[CrossProductMemberData(nameof(LoopbackServers), nameof(CookieNamesValues), nameof(TrueFalseBools))]
```
and then this would only need to define the member data for CookieNamesValues, using the shared LoopbackServers and TrueFalseBools functions that would be used in lots of places.

Reply:
Yeah, I think something like this would definitely be useful. 

If we are going to create our own MemberData attribute, then another useful thing would be to automatically map IEnumerable<T> to IEnumerable<object[]>, so that we don't need duplicative properties/methods for both the "typed" version and the "MemberData" version. E.g, "LoopbackServerFactories" and "LoopbackServerFactoriesMemberData" above. Instead of needing to do 

[MemberData(nameof(LoopbackServerFactoriesMemberData))]

you could just do

[SpecialMemberData(nameof(LoopbackServerFactories))]

and the SpecialMemberData logic would construct the object[]s for you. And if you specify multiple member data names, it would do the cross product thing. 

I'm not sure what to call this attribute though. Thoughts?